### ISSUE TITLE
Can Server `utils` be safely extracted to facilitate non-JAX-RS implementations

### ISSUE THREAD

**Issue by prkumar** [pingidentity/scim2#95]
Hi! First of all, Thanks for this amazing library!

I'm looking to use this library to implement a SCIM service on top of an existing Spring application. A lot of the tooling in `com.unboundid.scim2.server.utils` package (i.e., `ResourcePreparer`, `SchemaChecker`, etc.) seem really useful for my effort. However, it seems that to leverage these utilities I need to add an unnecessary dependency on JAX-RS. So, is there any way we can separate these server utilities from this constraint?

Please correct me if I'm wrong, but it seems that `ResourceTypeDefinition#fromJaxRsResource`, which relies on `javax.ws.rs.Path` annotation, is the the sole blocker here, as many other server utilities import the class, either directly or indirectly. Could we resolve this by moving the factory to a separate utility class (e.g., `ServerUtils`)?

**Comment by kqarryzada** [comment]
TL;DR: If you're using Spring, it's probably best to use `scim2-sdk-common` on its own and do a bit of the additional parsing manually.

&nbsp;

Hi @prkumar, thanks for sharing this suggestion. This is a good question, and I tried looking into a way to make this possible since unnecessary dependencies can be undesirable. Unfortunately, I'm not sure this is feasible. I think this would require significant changes to `scim2-sdk-server`. 

Our usages of JAX-RS are fairly integrated in the scim2-sdk-server library, and aren't isolated to a particular package. In fact, several of the classes in the utils package already have been using JAX-RS objects, and this has been the case since the initial releases of the UnboundID SCIM 2 SDK. This includes [one of the classes you listed](https://github.com/pingidentity/scim2/blob/b536175587b2e997be115756206bbae58e0f58a2/scim2-sdk-server/src/main/java/com/unboundid/scim2/server/utils/ResourcePreparer.java#L87), `ResourcePreparer`, which makes use of `UriInfo` objects from JAX-RS.

The utilities in the `scim2-sdk-server` component are designed to be helpful if you're already using JAX-RS to develop your SCIM service provider. However, if you're using another library like Spring, you probably don't need JAX-RS at all, so these utilities aren't as powerful. Given this, if you're using Spring to develop a SCIM service provider application, I honestly think it's better to leverage the `scim2-sdk-common` component and use the model classes provided there. This requires some manual work for some of the tasks that the `scim2-sdk-server` component would ordinarily do for you, but it should be feasible.

If there are any particular utility functions that are onerous to implement manually, please open a new issue for us to consider supporting it outside of JAX-RS. We could try to add some of this tooling to `scim2-sdk-server` in a way that wouldn't require you to also include JAX-RS in your project.

Ultimately, with the current architecture of the library, it would require significant effort to refactor `scim2-sdk-server` in a way that is not biased towards applications leveraging JAX-RS. I'd expect that most users of `scim2-sdk-server` are using JAX-RS already though, so any effort for us to deprecate/remove library methods to support this request would require that existing users do extra work when updating to a new version of the library.

For these reasons, we don't plan to make these changes at this time. Thanks very much for opening discussion on this!
